numpy/core/shape_base.py

```diff
@@ -275,3 +278,62 @@ def hstack(tup):
         return _nx.concatenate(arrs, 0)
     else:
         return _nx.concatenate(arrs, 1)
+
+def stack(arrays, axis=0):
+    """
+    Join a sequence of arrays along a new axis.
+
+    .. versionadded:: 1.10.0
+
+    Parameters
+    ----------
+    arrays : sequence of ndarrays
+        Each array must have the same shape.
+    axis : int, optional
+        The axis along which the arrays will be stacked.
+
+    Returns
+    -------
+    stacked : ndarray
+        The stacked array has one more dimension than the input arrays.
+
+    See Also
+    --------
+    concatenate : Join a sequence of arrays along an existing axis.
+    split : Split array into a list of multiple sub-arrays of equal size.
+
+    Examples
+    --------
+    >>> arrays = [np.random.randn(3, 4) for _ in range(10)]
+    >>> np.stack(arrays, axis=0).shape
+    (10, 3, 4)
+
+    >>> np.stack(arrays, axis=1).shape
+    (3, 10, 4)
+
+    >>> np.stack(arrays, axis=2).shape
+    (3, 4, 10)
+
+    >>> a = np.array([1, 2, 3])
+    >>> b = np.array([2, 3, 4])
+    >>> np.stack((a, b))
+    array([[1, 2, 3],
+           [2, 3, 4]])
+
+    >>> np.stack((a, b), axis=-1)
+    array([[1, 2],
+           [2, 3],
+           [3, 4]])
+
+    """
+    arrays = [_nx.asarray(arr) for arr in arrays]
```

Comment:
Do the existing *_stack functions perform broadcasting?
On Feb 25, 2015 12:17 PM, "Marten van Kerkwijk" <notifications@github.com>
wrote:

> In numpy/core/shape_base.py
> <https://github.com/numpy/numpy/pull/5605#discussion_r25374756>:
>
> > +    >>> np.stack(arrays, axis=2).shape
> > +    (3, 4, 10)
> > +
> > +    >>> a = np.array([1, 2, 3])
> > +    >>> b = np.array([2, 3, 4])
> > +    >>> np.stack((a, b))
> > +    array([[1, 2, 3],
> > +           [2, 3, 4]])
> > +
> > +    >>> np.stack((a, b), axis=-1)
> > +    array([[1, 2],
> > +           [2, 3],
> > +           [3, 4]])
> > +
> > +    """
> > +    arrays = [_nx.asarray(arr) for arr in arrays]
>
> Actually, an option may be to add a subok keyword argument, and then
> replace this line with:
>
> arrays = [_nx.array(arr, copy=False, subok=subok) for arr in arrays]
>
> (which should be slightly faster as it avoids the indirection of asarray)
>
> It does mean adding to the documentation that how subclasses get treated
> in the end also depends on concatenate. In that respect, a default of
> False for subok may be reasonable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5605/files#r25374756>.
>